Blog Post: Going beyond the Abstract Syntax Tree (AST) with the TypeScript Type Checker (#88)
satellytes/satellytes.com: blog-posts/typescript-ast-type-checker.md

---
path: "/blog/typescript-ast-type-checker"
date: "2021-06-18"
title: "Going beyond the Abstract Syntax Tree (AST) with the TypeScript Type Checker"
shortSummary: How to analyze a typescript file beyond AST (Abstract Syntax Tree) with the tpe checker.
author: Georgios Kaleadis
authorSummary: "CTO at Satellytes"
featuredImage: images/johann-siemens-EPy0gBJzzZU-unsplash.jpg
attribution:
    creator: Johann Siemens
    source: https://unsplash.com/photos/EPy0gBJzzZU

---

> For a recent Angular project we had to inspect a set of typescript files and output a table of involved types which are resolved down to their primitives. This post is about how to collect that data and why we need to go beyond the AST.

Reviewer: I missing the "Common usecase" for this. Why is this important in general? I guess it's for automatic type generation for a component documentation?

Author: A small update to clarify the purpose. Very good hint as I should clarify the purpose of the whole drama a little bit better 😄

> In a recent Angular typescript project we defined data contracts between components based on generics. Later the components are arranged and connected within a low-code platform which will check if two components can work together based on their data contract. The checking process involves a json dataset of typing information which we collect by recursively resolving typescript types down to their primitives and then store the found property names along with their typing name. This post is about how to collect the typing information and why we need to go beyond the AST to achieve this.

```typescript
// prints `collectedValue: ImportantValue`
console.log(`└── ${collectedValueProperty.name}: ${propertyTypeName}`)
```
What's left is to dive one level deeper to finally extract the types from the nested `ImportantValue`. This is basically *"rinse & repeat"* as you will see in the following code example. Instead of extracing the first element we use a for-loop though in order to find all properties.

```typescript
// remember we are now processing `ImportantValue` which is stored in `propertyType`
for (const nestedProperty of propertyType.getProperties()) {
  const nestedPropertyType = checker.getTypeOfSymbolAtLocation(nestedProperty, node);
  const nestedPropertyTypeName = checker.typeToString(nestedPropertyType);
  /** prints the following
   ├── value1: string
   ├── value2: number
   ├── value3: Date
   */
  console.log(`  ├── ${nestedProperty.name}: ${nestedPropertyTypeName}`)
}
```

<details>
    <summary>Full Source Example</summary>

>     ````typescript

Reviewer: One tick too much ;-)

Author: this is by purpose here, because there is a nested tick group and you can make the ticks arbitrarily long (so it's four tickers) to group nested ticks (with three ticks) 🤓 

````
A three ticks examples wrapped by four ticks to make it a code group:

```
works 👌
```
````

### Recursion

First, let's make the analysis recursive to find every property in any given file.

```typescript
function processProperty(type: ts.Type, node: ts.Node, level = 0) {
  if(level === 0) {
    console.group(`.\n└──Processing '${checker.typeToString(type)}'`)
  }

  for (const property of type.getProperties()) {
    const propertyType = checker.getTypeOfSymbolAtLocation(property, node);
    const propertyTypeName = checker.typeToString(propertyType);

    processProperty(propertyType, node, level + 1)
    console.log(`  ├── ${property.name}: ${propertyTypeName}`)

  }
  console.groupEnd();

}

ts.forEachChild(myComponentSourceFile, node => {
  if (ts.isTypeAliasDeclaration(node) && node.name.escapedText === "Output") {
    const outputType = checker.getTypeAtLocation(node.name);
    processProperty(outputType, node);
  }
});
```

This will find every single property, no matter how deep it's nested. That's because `processProperty()` is used recursively on all nested properties.

When you run this code, you will be lost in noise. See the following log and try to spot our types within the ocean of properties pouring in from the standard library.

> <details>

Reviewer: `margin-bottom` is missing here. We could fix this separately.

<img width="909" alt="Screenshot 2021-06-21 at 08 02 46" src="https://user-images.githubusercontent.com/7814926/122714183-33f39d00-d267-11eb-9931-06d56c7c3182.png">

Author: I added 16px but we still need to check how this behaves in detail.

<summary>Output</summary>

>     ```typescript

Reviewer: Hm, the code is not formatted here although the syntax is correct 🤔

Author: The ticks are not allowed to be bested. I pulled them to the front now everything is highlighted.

      ├── setUTCDate: (date: number) => number
      ├── setMonth: (month: number, date?: number) => number
      ├── setUTCMonth: (month: number, date?: number) => number
      ├── setFullYear: (year: number, month?: number, date?: number) => number
      ├── setUTCFullYear: (year: number, month?: number, date?: number) => number
      ├── toUTCString: () => string
      ├── toISOString: () => string
      ├── toJSON: (key?: any) => string
      ├── getVarDate: () => VarDate
      ├── __@toPrimitive@755: { (hint: "default"): string; (hint: "string"): string; (hint: "number"): number; (hint: string): string | number; }
    👉├── value3: Date
    👉├── collectedValue: ImportantValue
    ```
</details>

That's the "standard library" issue described earlier. The `Date` and `string` types causes this drama and we need to stop our processing before entering those types.

### Exclude the standard types

TypeScript gives us plenty of tools to do that. Here is a helper method `isTypeLocal` I have build for our use cases.

```typescript
function isTypeLocal(symbol: ts.Symbol) {
  const sourceFile = symbol?.valueDeclaration?.getSourceFile();
  const hasSource = !!sourceFile;

>   const isStandardLibrary = hasSource && program.isSourceFileDefaultLibrary(sourceFile!)

Reviewer: Is `isSourceFileDefaultLibrary` also a build in method? Wow, would be cool to have a small comment on how Typescript decides which library is default.

Reviewer: I just see you explain it a bit in the next sentence 😄 A bit more detailed would be very interesting, like if a default library is everything the comes with Typescript.

Author: Hey I'm a consumer not a typescript contributor so I have no clue 😅 But a standard library is very well defined through this part in the tsconfig.json which should look familiar.

```
{
  "lib": ["es5", "es2015"]
}
```

I guess every file imported from there is marked with a flag to signal the standard library membership which can be queries through the inversion `file.hasNoDefaultLib`.

I added a link to the tsconfig lib documentation but I would like to focus on the API and chose not to dive deeper than necessary, although the entire TS processing is so damn interesting 🤩

Resources:
+ https://www.typescriptlang.org/tsconfig#lib 
+ https://github.com/microsoft/TypeScript/tree/main/lib
+ https://github.com/microsoft/TypeScript/blob/5afe42e14e61d7e4df5d75cc0022283711cb593a/src/compiler/program.ts#L1721-L1739

  const isExternal = hasSource && program.isSourceFileFromExternalLibrary(sourceFile!);
  const hasDeclaration = !!symbol?.declarations?.[0];

  return !(isStandardLibrary || isExternal) && hasDeclaration;
}
```

The method will detect if a given symbol belongs to a standard library (`Date`), to an external library (whatever you use from `node_modules`) and everything that doesn't have an actual declaration like primitive types (`string`, `number`).

We will use that helper to prevent our recursion from branching into those unwanted types:

```typescript
if(isTypeLocal(propertySymbol)) {
  // It's a type we have defined, so print it
  // and then process its nested properties
  console.group(`  └── ${property.name}: ${propertyTypeName}`)
  processProperty(propertyType, node, level + 1)
}else {
  // It's not a local type, so print it but don't do anything further
  console.log(`  ├── ${property.name}: ${propertyTypeName}`)
}
```

The updated code can process the initial file but it's much more flexible. Let's process a much deeper nested type `Output` and watch the console.

<details>

>     <summary>Updated file `file-with-types.ts`</summary>

Reviewer: Not formatted. Also a candidate for a separate PR

<img width="434" alt="Screenshot 2021-06-21 at 08 09 24" src="https://user-images.githubusercontent.com/7814926/122714692-022f0600-d268-11eb-86f4-6cac095d08be.png">

Author: yes fixed with the correct indentation.

    source: https://unsplash.com/photos/EPy0gBJzzZU

---

We currently develop a [low-code platform](https://en.wikipedia.org/wiki/Low-code_development_platform) for an enterprise client where Angular components are arranged and connected based on a given configuration file. The components define *data contracts* based on [generics](https://www.typescriptlang.org/docs/handbook/2/generics.html) so the platform always knows what data can flow between them. In order to perform the actual type check we use a json file that already contains all relevant typing information. We collect the typing data by recursively resolving typescript types down to their primitives and then store the found property names along with their typing name. This post is about how to collect the typing information and why we need to go beyond the AST to achieve this.

<!-- stop excerpt -->

You can find the repository on [github.com/georgiee/typescript-type-checker-beyond-ast](https://github.com/georgiee/typescript-type-checker-beyond-ast) and you can directly run the given example in your browser with [code sandbox](https://githubbox.com/georgiee/typescript-type-checker-beyond-ast)

## Expectations

> Look at the two types below. You can find primitives like `string` and `number`, types from the standard library such as `Date` and type aliases like `NestedObjectType` that refer to object types which are assembled types that can contain primitives and other object types.

Reviewer 2: Typo
```suggestion
Look at the two types below. You can find primitives like `string` and `number`, types from the standard library such as `Date` and type aliases like `NestedObjectType` that refers to object types which are assembled types that can contain primitives and other object types.
```

Author: 🙏

Let's find out how we can approach this problem.

## Can we use the AST?
The [AST (Abstract Syntax Tree)](https://en.wikipedia.org/wiki/Abstract_syntax_tree) quickly comes to your mind to approach this problem. The AST is a data structure to represent the structure of your source file in a format readable by machines. Indeed, if I throw the above example in the [TypeScript AST Viewer](https://ts-ast-viewer.com) I get immediate access to the AST. 

![](images/ts-ast-viewer.png)

That output looks promising. I guess this could work for very simple types 👍

The problem with the AST: it's a static analysis, which means you're processing code without executing it. That's why you are missing information from the runtime. Typescript needs to run the code to understand it and to add additional semantics. You will encounter the following problems when you try to approach the problem with the AST:

+ The AST can't see imported files as `import` statements are not processed

> + [Created types](https://www.typescriptlang.org/docs/handbook/2/types-from-types.html) with operands like `keyof`& `typeof` are constructed only during runtime.

Reviewer 2: Add missing space and no need to add a dot at the end if there is only one sentence (matter of taste, but at least it should be the same across other occurences).
```suggestion
+ [Created types](https://www.typescriptlang.org/docs/handbook/2/types-from-types.html) with operands like `keyof` & `typeof` are constructed only during runtime
```

Author: thanks 🙏

> + [Advanced types](https://www.typescriptlang.org/docs/handbook/advanced-types.html) like [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types) or [conditional types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#conditional-types) also rely on being processed by typescript otherwise you have no chance to understand and process them.

Reviewer 2: Advanced types page is marked as deprecated and a successor linked, applies to all three links.
Not 100% sure if the updated links reflect exactly the same.

```suggestion
+ [Advanced types](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) like [type guards](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) or [conditional types](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#equality-narrowing) also rely on being processed by typescript otherwise you have no chance to understand and process them
```

Author: Thanks, I have updated the links and I will dismiss this suggestion. I don't know how I missed that big deprecation banner 😄

> This means, the AST approach is a dead end.

Reviewer 2: Suggestion to avoid to many _this, that, the_ line beginnings.
```suggestion
Eventually, the AST approach is a dead end.
```

Author: thanks, I see and feel this but I'm lacking a suitable pool of words that I have at hands while writing 🙏

The following code shows the most crucial parts of type introspection with TS. Create a program, derive the checker and then use that checker for your analysis.

```typescript

const program: ts.Program = ts.createProgram(files, tsConfig);
const checker: ts.TypeChecker = program.getTypeChecker();

// Later involve the checker somehow 
const classSymbol = checker.getSymbolAtLocation(node.name);
// ...
```

## Type Checker Usage
Getting the checker setup is pretty straightforward, as usually it gets complicated with all the details. Let's tackle it step by step. We start by preparing a file `file-with-types.ts` that should contain the types we want to examine.

```typescript
// file-with-types.ts

type MainObjectType = {
  propertyWithTypeAlias: NestedObjectType;
};

type NestedObjectType = {
  value1: string;
  value2: number;
  value3: Date;
};
```

Together with this file, we want to answer the following question:

> > Using `checker.ts`, how can we access the details of the type `NestedObjectType` so we know that the proeprty `propertyWithTypeAlias` on `MainObjectType` has three distinctive nested properties?

Reviewer 2: Typo
```suggestion
> Using `checker.ts`, how can we access the details of the type `NestedObjectType` so we know that the property `propertyWithTypeAlias` on `MainObjectType` has three distinctive nested properties?
```

  ├── setUTCMinutes: (min: number, sec?: number, ms?: number) => number
  ├── setHours: (hours: number, min?: number, sec?: number, ms?: number) => number
  ├── setUTCHours: (hours: number, min?: number, sec?: number, ms?: number) => number
  ├── setDate: (date: number) => number
  ├── setUTCDate: (date: number) => number
  ├── setMonth: (month: number, date?: number) => number
  ├── setUTCMonth: (month: number, date?: number) => number
  ├── setFullYear: (year: number, month?: number, date?: number) => number
  ├── setUTCFullYear: (year: number, month?: number, date?: number) => number
  ├── toUTCString: () => string
  ├── toISOString: () => string
  ├── toJSON: (key?: any) => string
  ├── getVarDate: () => VarDate
  ├── __@toPrimitive@755: { (hint: "default"): string; (hint: "string"): string; (hint: "number"): number; (hint: string): string | number; }
👉├── value3: Date
👉├── propertyWithTypeAlias: NestedObjectType
```
</details>

That's the "standard library" issue described earlier. The `Date` and `string` types causes this drama and we need to stop our processing before entering those types.

### Exclude the standard types

> TypeScript gives us plenty of tools to do that. Here is a helper method `isTypeLocal` I have build for our use cases.

Reviewer 2: Typo
```suggestion
TypeScript gives us plenty of tools to do that. Here is a helper method `isTypeLocal` I have built for our use cases.
```
